lib/codecs/src/decoding/format/avro.rs

```diff
@@ -0,0 +1,302 @@
+            log.insert(event_path!(k.as_str()), try_from(v)?);
+        }
+        Ok(log)
+    }
+}
+
+impl Deserializer for AvroDeserializer {
+    fn parse(
+        &self,
+        bytes: Bytes,
+        log_namespace: LogNamespace,
+    ) -> vector_common::Result<SmallVec<[Event; 1]>> {
+        let log = self.parse_single(bytes, log_namespace)?;
+        Ok(smallvec![log.into()])
+    }
+}
+
+// can't use std::convert::TryFrom because of orphan rules
+pub fn try_from(value: AvroValue) -> vector_common::Result<VrlValue> {
+    // very similar to avro to json see `impl std::convert::TryFrom<AvroValue> for serde_json::Value`
+    // LogEvent has native support for bytes, so it is used for Bytes and Fixed
+    match value {
+        AvroValue::Array(array) => {
+            let mut vector = Vec::new();
+            for item in array {
+                vector.push(try_from(item)?);
+            }
+            Ok(VrlValue::Array(vector))
+        }
+        AvroValue::Boolean(boolean) => Ok(VrlValue::from(boolean)),
+        AvroValue::Bytes(bytes) => Ok(VrlValue::from(bytes)),
+        AvroValue::Date(d) => Ok(VrlValue::from(d)),
+        AvroValue::Decimal(ref d) => Ok(<Vec<u8>>::try_from(d)
+            .map(|vec| VrlValue::Array(vec.into_iter().map(VrlValue::from).collect()))?),
+        AvroValue::Double(double) => Ok(VrlValue::from(double)),
+        AvroValue::Duration(d) => Ok(VrlValue::Array(
+            <[u8; 12]>::from(d)
+                .into_iter()
+                .map(VrlValue::from)
+                .collect(),
+        )),
+        AvroValue::Enum(_, string) => Ok(VrlValue::from(string)),
+        AvroValue::Fixed(_, bytes) => Ok(VrlValue::from(bytes)),
+        AvroValue::Float(float) => Ok(VrlValue::from(float as f64)),
+        AvroValue::Int(int) => Ok(VrlValue::from(int)),
+        AvroValue::Long(long) => Ok(VrlValue::from(long)),
+        AvroValue::Map(items) => items
+            .into_iter()
+            .map(|(key, value)| try_from(value).map(|v| (key, v)))
+            .collect::<Result<Vec<_>, _>>()
+            .map(|v| VrlValue::Object(v.into_iter().collect())),
+        AvroValue::Null => Ok(VrlValue::Null),
+        AvroValue::Record(items) => items
+            .into_iter()
+            .map(|(key, value)| try_from(value).map(|v| (key, v)))
+            .collect::<Result<Vec<_>, _>>()
+            .map(|v| VrlValue::Object(v.into_iter().collect())),
+        AvroValue::String(string) => Ok(VrlValue::from(string)),
+        AvroValue::TimeMicros(time_micros) => Ok(VrlValue::from(time_micros)),
+        AvroValue::TimeMillis(time_millis) => Ok(VrlValue::from(time_millis)),
+        AvroValue::TimestampMicros(timestamp_micros) => Ok(VrlValue::from(timestamp_micros)),
+        AvroValue::TimestampMillis(timestamp_millis) => Ok(VrlValue::from(timestamp_millis)),
+        AvroValue::Union(_, v) => try_from(*v),
+        AvroValue::Uuid(uuid) => Ok(VrlValue::from(uuid.as_hyphenated().to_string())),
```

Comment:
💬 We should treat this as a `Vec<u8>>`. Let's also add UUID in a test.

Reply:
In the avro doc, [the uuid is a logic type annotates an string](https://avro.apache.org/docs/1.11.1/specification/#uuid), there maybe some trouble when avro encoding uuid if we treat uuid as `Vec<u8>`.

Reply:
💬 Do we need the uuid hyphenated as well? This is the type of thing a roundtrip test would catch. See my suggestion bellow.

Reply:
I tried the following code which try to serialize the uuid data to avro format. If we use type `uuid` or `String` for uuid data, the encoding will success, but `Vec<u8>` will fail. So I think that let uuid data stay `String` format is better than treating uuid as Vec<u8>.

```rust
let raw_schema = r#"
{
    "type": "string",
    "logicalType":"uuid"
}
"#;
let schema = Schema::parse_str(raw_schema).unwrap();

// This branch success
let test = uuid!("67e55044-10b1-426f-9247-bb680e5fe0c8");
let test:String = test.as_hyphenated().to_string();
let value = apache_avro::to_value(test).unwrap();
let value = value.resolve(&schema).unwrap();
let bytes = apache_avro::to_avro_datum(&schema, value).unwrap();
println!("{:?}", bytes);

// This branch success
let test:Uuid = uuid!("67e55044-10b1-426f-9247-bb680e5fe0c8");
let value = apache_avro::to_value(test).unwrap();
let value = value.resolve(&schema).unwrap();
let bytes = apache_avro::to_avro_datum(&schema, value).unwrap();
println!("{:?}", bytes);

// This branch will faill
let test = uuid!("67e55044-10b1-426f-9247-bb680e5fe0c8");
let test:Vec<u8> = test.as_bytes().to_vec();
let value = apache_avro::to_value(test).unwrap();
let value = value.resolve(&schema).unwrap();
let bytes = apache_avro::to_avro_datum(&schema, value).unwrap();
println!("{:?}", bytes);
```

Reply:
This makes sense, thank you for double checking. We should include this in the tests. 
In general this PR looks good now! There are a few unresolved comments mainly about strengthening tests and then we can merge.

Reply:
Added a test case, please review and fix if you found any bugs.